db/write_thread.cc

```diff
@@ -310,137 +356,295 @@ size_t WriteThreadImpl::EnterAsBatchGroupLeader(
-    CreateMissingNewerLinks(head);
-    assert(last_writer->link_newer->link_older == last_writer);
-    last_writer->link_newer->link_older = nullptr;
-
-    // Next leader didn't self-identify, because newest_writer_ wasn't
-    // nullptr when they enqueued (we were definitely enqueued before them
-    // and are still in the list).  That means leader handoff occurs when
-    // we call MarkJoined
-    SetState(last_writer->link_newer, STATE_GROUP_LEADER);
-  }
-  // else nobody else was waiting, although there might already be a new
-  // leader now
-
-  while (last_writer != leader) {
-    last_writer->status = status;
-    // we need to read link_older before calling SetState, because as soon
-    // as it is marked committed the other thread's Await may return and
-    // deallocate the Writer.
-    auto next = last_writer->link_older;
-    SetState(last_writer, STATE_COMPLETED);
-
-    last_writer = next;
+  if (enable_pipelined_write_) {
+    Writer* next_leader = nullptr;
+    // We attempt to put a dummy writer to the head of newest_writer_ list
+    // with a CAS operation, to block the next batch group leader.
+    // If the CAS operation fail, we know the next leader is in the list at
+    // this point. We simply don't wake it up until we are ready.
+    Writer dummy;
+    Writer* newest_writer = last_writer;
+    bool has_dummy =
+        newest_writer_.compare_exchange_strong(newest_writer, &dummy);
+    if (!has_dummy) {
+      CreateMissingNewerLinks(newest_writer);
+      next_leader = last_writer->link_newer;
+      assert(next_leader != nullptr);
+    }
+    // Notify writers don't write to memtable to exit.
+    for (Writer* w = last_writer; w != leader;) {
+      Writer* next = w->link_older;
+      w->status = status;
+      if (!w->ShouldWriteToMemtable()) {
+        CompleteFollower(w, write_group);
+      }
+      w = next;
+    }
+    if (!leader->ShouldWriteToMemtable()) {
+      CompleteLeader(write_group);
+    }
+    // Link the ramaining of the group to memtable writer list.
+    if (write_group.size > 0) {
+      if (LinkGroup(write_group, newest_memtable_writer_)) {
+        // The leader can now be different from current writer.
+        SetState(write_group.leader, STATE_MEMTABLE_WRITER_LEADER);
+      }
+    }
+    // Remove the dummy writer. If the CAS operation fail, the next leader is
+    // next to the dummy writer in the list.
+    if (has_dummy) {
+      newest_writer = &dummy;
+      if (!newest_writer_.compare_exchange_strong(newest_writer, nullptr)) {
+        CreateMissingNewerLinks(newest_writer);
+        next_leader = dummy.link_newer;
+        assert(next_leader != nullptr);
+      }
+    }
+    // Wake up the next leader.
+    if (next_leader != nullptr) {
+      next_leader->link_older = nullptr;
+      SetState(next_leader, STATE_GROUP_LEADER);
+    }
+    AwaitState(leader, STATE_MEMTABLE_WRITER_LEADER |
+                           STATE_PARALLEL_MEMTABLE_WRITER | STATE_COMPLETED,
```

Comment:
The indentation is done by clang-format. Apparently it align this line with `STATE_MEMTABLE_WRITER_LEADER ` and add two indents.